src/components/Ledger/LedgerConnect.js

```diff
@@ -0,0 +1,377 @@
+    error: null,
+    refreshing: false,
+    waiting: false,
+  }
+
+  _subscriptions: ?{unsubscribe: () => any} = null
+  _bluetoothEnabled: ?boolean = null
+  _transportLib: Object = null
+  _isMounted: boolean = false
+
+  async componentDidMount() {
+    const {useUSB} = this.props
+    this._transportLib = useUSB === true ? TransportHID : TransportBLE
+    this._isMounted = true
+    if (useUSB === false) {
+      if (Platform.OS === 'android') {
+        await PermissionsAndroid.request(
+          PermissionsAndroid.PERMISSIONS.ACCESS_FINE_LOCATION,
+        )
+      }
+      // check if bluetooth is available
+      // no need to save a reference to this subscription's unsubscribe func
+      // as it's just an empty method. Rather, we make sure sate is only
+      // modified when component is mounted
+      let previousAvailable = false
+      TransportBLE.observeState({
+        next: (e) => {
+          if (this._isMounted) {
+            Logger.debug('BLE observeState event', e)
+            if (this._bluetoothEnabled == null && !e.available) {
+              this.setState({
+                error: new BluetoothDisabledError(),
+                refreshing: false,
+              })
+            }
+            if (e.available !== previousAvailable) {
+              previousAvailable = e.available
+              this._bluetoothEnabled = e.available
+              if (e.available) {
+                this.reload()
+              } else {
+                this.setState({
+                  error: new BluetoothDisabledError(),
+                  refreshing: false,
+                  devices: [],
+                })
+              }
+            }
+          }
+        },
+      })
+    }
+    this.startScan()
+  }
+
+  componentWillUnmount() {
+    this._unsubscribe()
+    this._isMounted = false
+  }
+
+  startScan = () => {
+    const {useUSB} = this.props
+    this.setState({refreshing: true})
+
+    this._subscriptions = this._transportLib.listen({
+      complete: () => {
+        Logger.debug('listen: subscription completed')
+        this.setState({refreshing: false})
+      },
+      next: (e) => {
+        if (e.type === 'add') {
+          Logger.debug('listen: new device detected')
+          if (useUSB === true) {
+            // if a device is detected, save it in state immediately
+            this.setState({
+              refreshing: false,
+              deviceObj: e.descriptor,
+            })
+          } else {
+            // with bluetooth, new devices are appended in the screen
+            this.setState(deviceAddition(e.descriptor))
+          }
+        }
+      },
+      error: (error) => {
+        this.setState({error, refreshing: false, devices: []})
+      },
+    })
+  }
+
+  _unsubscribe: () => void = () => {
+    if (this._subscriptions != null) {
+      this._subscriptions.unsubscribe()
+      this._subscriptions = null
+    }
+  }
+
+  _setStateSafe: ($Shape<State>) => void = (newState) => {
```

Comment:
FYI `$Shape` is deprecated. If you want to get an inexact subset of the State, consider something like this utility type

```
declare type InexactSubset<T: {...}> = $Rest<T, {...}>;
```